File: OperatingSystems/assign4/assign4.java
import java.util.*;

    class SelectionSort extends Thread {
        private double [] dArr;
        private int iStartingV;
        private int iEndingV;

        public SelectionSort(double [] arr, int startingV, int endingV) {
            dArr = arr;
            iStartingV = startingV;
            iEndingV = endingV;
        }

        public void run() {
            int i; 
            int j; 
            int min;
            double tmp;
            dArr = dArr;
            for(i = iStartingV; i < iEndingV; i++) {
                min = i;
                for(j = i + 1; j < iEndingV; j++) {
                    if(dArr[j] < dArr[min])
                        min = j;
                }
                swap(dArr[i],dArr[min]);
            }
        }
        
        public void swap(double dpArrA, double dpArrB) {
            double tmp;
            tmp = dpArrA;
            dpArrA = dpArrB;
            dpArrB = tmp;
        }
    };
    
	    class merge extends Thread {
        private double [] dInArr; 
        private double [] dOutArr;
        private int iMid;

        public merge(double [] iArr, double [] oArr, int m) {
            dInArr = iArr;
            dOutArr = oArr;
            iMid = m;
        }

        public void run() {
            int i = 0;
            int j;
            int k = 0;
            j = iMid;
            while(i < iMid && j < dInArr.length) {
                if(dInArr[i] < dInArr[j])
                    dOutArr[k++] = dInArr[i++];
                else
                    dOutArr[k++] = dInArr[j++];
            }
            while(i < iMid)
                dOutArr[k++] = dInArr[i++];
            while(j < dInArr.length)
                dOutArr[k++] = dInArr[j++];
        }
    };
	
public class assign4 {

    public static void main(String []arg) {
        double [] aArr; //A
        double [] bArr; //B
        double [] cArr; //C
        int mid;
        int iNput;
        int i;
        long start;
        long end;

        if(arg.length != 1) {
            System.out.println("ERROR USAGE: java assign4 (Integer)");
            return;
        }

        //iNput will be given as command line argument
        iNput = Integer.parseInt(arg[0]);
        
        //create array A ( n double values) and randomly generate these values 
        aArr = new double[iNput];
        
        //B same size of A
        bArr = new double[iNput];
        
        //C same size of A
        cArr = new double[iNput];
        
        
        Random rand = new Random(System.currentTimeMillis());

        for(i = 0; i < iNput; i++) { 
            aArr[i] = rand.nextDouble() * 100.0; 
        }

        //Create threads
        //Afirsthalf
        SelectionSort thA1 = new SelectionSort(aArr, 0, iNput/2);
        
        //AsecondHalf
        SelectionSort thA2 = new SelectionSort(aArr, iNput/2, aArr.length);
        
        //copy A into B
        for(i = 0; i < iNput; i++) { 
            aArr[i] = bArr[i];
        }
        
        //create thM  mergeThread to merge Afirsthalf and Asecondhalf into C
        merge thM = new merge(aArr, cArr, iNput/2);
        
        //doing two threads first for the sake of the example output
        //start = clock();
        start = System.nanoTime(); 

        thA1.start();
        thA2.start();

        try {
            //join thA1
            thA1.join();
            //join thA2
            thA2.join();
        } catch (Exception e) {
            e.printStackTrace();
        }

        //create thM  mergeThread to merge Afirsthalf and Asecondhalf into C
        thM.start();
        try {
            //join thM
            thM.join();
        } catch (Exception e) {
            e.printStackTrace();
        }
        
        //Sorting is done in (end - start) ms when two threads are used 
        //End = clock() - start;
        end = System.nanoTime() - start; 

        System.out.println("Sorting is done in " + end/1000000.0f + "ms when two threads are used");
       
        //Time for single thread
         //create thB sortThread to sort B
        SelectionSort thB = new SelectionSort(bArr, 0, iNput);
        start = System.nanoTime(); 
        
        //start = clock();
        thB.start();
        try {
            thB.join();
        } catch (Exception e) {
            e.printStackTrace();
        }
        
        //end = clock() - start;
        end = System.nanoTime() - start; 

        System.out.println("Sorting is done in "+ (end/1000000.0f) + "ms when one thread is used");
    }
}

File: OperatingSystems/assign4/assign4.c
#include <fcntl.h>
#include <stdio.h>
#include <stdlib.h>
#include <time.h>
#include <unistd.h>

//array for sorting!
struct sArray {
    double *dpArr;
	int iInput;
    int iStartingV;
    int iEndingV;
};

//array for merging!
struct mArray {
    double *dInArr; 
    double *dOutArr;
    int iInput;
    int iMid;
};

int swap(double dpArrA, double dpArrB);
int selectionSort(void *ptr);
int merge(void *ptr);

int main(int argc, char *argv[]) {
clock_t end;
clock_t start;
double * aArr; //A
double * bArr; //B
double * cArr; //C
int i;
int iMid;
int iNput;
pthread_t thA1; //thA1
pthread_t thA2; //thA2
pthread_t thB;  //thB
pthread_t thM; //thM
struct sArray aFirstHalf;
struct sArray aSecondHalf;
struct mArray mThread;

if(argc != 2) {
	//error check number of args
      printf("ERROR USAGE: %s (Integer) \n", argv[0]);
      return -1;
}

//iNput will be given as command line argument
iNput = atoi(argv[1]);

//create array A ( n double values) and randomly generate these values 
aArr = (double*)malloc(iNput * sizeof(double));
//B same size of A
bArr = (double*)malloc(iNput * sizeof(double));
//C same size of A
cArr = (double*)malloc(iNput * sizeof(double));

srand(time(NULL));

for(i = 0; i < iNput; i++) {
    aArr[i] = (double)rand() / RAND_MAX * 99.0 + 1.0; /* range 1.0 - 100.0 */
}

//Afirsthalf  and AsecondHalf with the half size of A

//copy A into Afirsthalf  and AsecondHalf

//Afirsthalf
aFirstHalf.iStartingV = 0;
aFirstHalf.iEndingV = iNput/2;
aFirstHalf.dpArr = aArr;

//AsecondHalf
aSecondHalf.iStartingV = iNput/2;
aSecondHalf.iEndingV = iNput; //iNput being TOTAL value
aSecondHalf.dpArr = aArr;

//copy A into B
for(i = 0; i < iNput; i++) {
	aArr[i] = bArr[i];
}


//create thM  mergeThread to merge Afirsthalf and Asecondhalf into C
mThread.iMid = iNput/2;
mThread.iInput = iNput;
mThread.dInArr = aArr;
mThread.dOutArr = cArr;

//doing two threads first for the sake of the example output

start = clock();

//int pthread_create(pthread_t *thread, const pthread_attr_t *attr,
//                        void *(*start_routine) (void *), void *arg);

//create thA1  sortThread to sort Afirsthalf   /* use selection/insertion sort O((n/2)^2) */
pthread_create(&thA1, NULL, selectionSort, &aFirstHalf);
//create thA2  sortThread to sort Asecondhalf    /* use selection/insertion sort O((n/2)^2) */
pthread_create(&thA2, NULL, selectionSort, &aSecondHalf);


//join thA1
//int pthread_join(pthread_t thread, void **retval);
pthread_join(thA1, NULL);
//join thA2
pthread_join(thA2, NULL);

//create thM  mergeThread to merge Afirsthalf and Asecondhalf into C

pthread_create(&thM, NULL, merge, &mThread);
//join thM
pthread_join(thM, NULL);

//Sorting is done in (end - start) ms when two threads are used 
end = clock() - start;

printf("Sorting is done in %f ms when two threads are used\n", end * 1000.0f / CLOCKS_PER_SEC);

//Time for single thread
aFirstHalf.iEndingV = iNput;
aFirstHalf.dpArr = bArr; 
start = clock();

//create thB sortThread to sort B
pthread_create(&thB, NULL, selectionSort, &aFirstHalf);
selectionSort(&aFirstHalf);
pthread_join(thB);


//Sorting is done in (end - start) ms when one thread is used 
end = clock() - start; 

printf("Sorting is done in %f ms when one thread is used\n", end * 1000.0f / CLOCKS_PER_SEC);

free(aArr);
free(bArr);
free(cArr);
return 0;
}

int selectionSort(void *ptr) {
int i;
int j;
int minIdx;
double tmp;
double *dpArr;
struct sArray* p = (struct sArray*) ptr;
dpArr = p->dpArr;
for(i = p->iStartingV; i < p->iEndingV; i++) {
    minIdx = i;
    for(j = i + 1; j < p->iEndingV; j++) {
      if(dpArr[j] < dpArr[minIdx])
        minIdx = j;
    }
	//Swap
	swap(dpArr[i], dpArr[minIdx]);
}
return 0;
}

int swap(double dpArrA, double dpArrB) {
	double tmp = dpArrA;
	dpArrA = dpArrB;
	dpArrB = tmp;
	return 0;
}

int merge(void *ptr) {
double *in;
double *out;
int i = 0;
int j;
int k = 0;
struct mArray* p = (struct mArray*) ptr;
j = p->iMid;
in = p->dInArr;
out = p->dOutArr;
while(i < p->iMid && j < p->iInput) {
    if(in[i] < in[j])
      out[k++] = in[i++];
    else
      out[k++] = in[j++];
}
while(i < p->iMid)
    out[k++] = in[i++];
while(j < p->iInput)
    out[k++] = in[j++];
return 0;
}